### ISSUE TITLE
Management Repair failed with error repair status giving up

### ISSUE THREAD

**Issue by aleksbykov** [scylladb/scylla-manager#3584]
Job: https://jenkins.scylladb.com/job/scylla-master/job/scylla-master-perf-regression-latency-650gb-with-nemesis/46/
TestId: c9190059-48e6-4310-a49f-70c41796b9d7
Scylla version: 5.4.0~dev.20230914.c27d212f4bc2 with build-id af2c3872b1f1d4f59a8fa93c9d9bcd92e3971418
Scylla manager: 3.2.0-0.20230815.9db309c5

DB logs: https://cloudius-jenkins-test.s3.amazonaws.com/c9190059-48e6-4310-a49f-70c41796b9d7/20230917_011800/db-cluster-c9190059.tar.gz
Manager log: 
[scylla_manager.tar.gz](https://github.com/scylladb/scylla-manager/files/12700825/scylla_manager.tar.gz)

During performance test_write, first operation was manager_repair. Manager repair start at 08:32, but after 4 hours it failed with next error:
```
Sep 16 12:38:26 perf-latency-nemesis-ubuntu-monitor-node-c9190059-1 scylla-manager[31169]: {"L":"ERROR","T":"2023-09-16T12:38:26.026Z","N":"repair","M":"Repair failed","error":"master 10.4.0.73 keyspace keyspace1 table standard1 command 13: get repair status: giving up after 10 attempts: after 1h0m0s: context deadline exceeded","_trace_id":"V0mA2RqoQUK17cRxEz01XQ","errorStack":"github.com/scylladb/scylla-manager/v3/pkg/service/repair.(*worker).runRepair.func1\n\tgithub.com/scylladb/scylla-manager/v3/pkg/service/repair/worker.go:78\ngithub.com/scylladb/scylla-manager/v3/pkg/service/repair.(*worker).runRepair\n\tgithub.com/scylladb/scylla-manager/v3/pkg/service/repair/worker.go:97\ngithub.com/scylladb/scylla-manager/v3/pkg/service/repair.(*worker).Run\n\tgithub.com/scylladb/scylla-manager/v3/pkg/service/repair/worker.go:48\ngithub.com/scylladb/scylla-manager/v3/pkg/service/repair.(*Service).Repair.func1\n\tgithub.com/scylladb/scylla-manager/v3/pkg/service/repair/service.go:319\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\tgolang.org/x/sync@v0.2.0/errgroup/errgroup.go:75\nruntime.goexit\n\truntime/asm_amd64.s:1598\n","S":"github.com/scylladb/go-log.Logger.log\n\tgithub.com/scylladb/go-log@v0.0.7/logger.go:101\ngithub.com/scylladb/go-log.Logger.Error\n\tgithub.com/scylladb/go-log@v0.0.7/logger.go:84\ngithub.com/scylladb/scylla-manager/v3/pkg/service/repair.(*generator).processResult\n\tgithub.com/scylladb/scylla-manager/v3/pkg/service/repair/generator.go:219\ngithub.com/scylladb/scylla-manager/v3/pkg/service/repair.(*generator).Run\n\tgithub.com/scylladb/scylla-manager/v3/pkg/service/repair/generator.go:191\ngithub.com/scylladb/scylla-manager/v3/pkg/service/repair.(*Service).Repair.func2\n\tgithub.com/scylladb/scylla-manager/v3/pkg/service/repair/service.go:324\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\tgolang.org/x/sync@v0.2.0/errgroup/errgroup.go:75"}
Sep 16 12:38:26 perf-latency-nemesis-ubuntu-monitor-node-c9190059-1 scylla-manager[31169]: {"L":"INFO","T":"2023-09-16T12:38:26.029Z","N":"repair","M":"Progress","percent":100,"count":6912,"success":6144,"failed":768,"_trace_id":"V0mA2RqoQUK17cRxEz01XQ"}
Sep 16 12:38:26 perf-latency-nemesis-ubuntu-monitor-node-c9190059-1 scylla-manager[31169]: {"L":"INFO","T":"2023-09-16T12:38:26.029Z","N":"repair","M":"Close generator","_trace_id":"V0mA2RqoQUK17cRxEz01XQ"}
Sep 16 12:38:26 perf-latency-nemesis-ubuntu-monitor-node-c9190059-1 scylla-manager[31169]: {"L":"INFO","T":"2023-09-16T12:38:26.029Z","N":"repair.worker 0","M":"Done","_trace_id":"V0mA2RqoQUK17cRxEz01XQ"}
Sep 16 12:38:26 perf-latency-nemesis-ubuntu-monitor-node-c9190059-1 scylla-manager[31169]: {"L":"INFO","T":"2023-09-16T12:38:26.042Z","N":"scheduler.00966dbf","M":"Run ended with ERROR","task":"repair/d7f016a9-1d5c-4c1e-9ec8-f76793512c32","status":"ERROR","cause":"repair 768 token ranges out of 6912","duration":"4h6m14.801874991s","_trace_id":"V0mA2RqoQUK17cRxEz01XQ"}
Sep 16 12:38:26 perf-latency-nemesis-ubuntu-monitor-node-c9190059-1 scylla-manager[31169]: {"L":"INFO","T":"2023-09-16T12:38:26.043Z","N":"scheduler.00966dbf","M":"Retry backoff","task":"repair/d7f016a9-1d5c-4c1
```
At these moment cluster consisted from 3 nodes. There is no any errors on node1, node2, node3.

**Comment by aleksbykov** [comment]
@ShlomiBalalis , can you take a look and help with it

**Comment by Michal-Leszczynski** [comment]
From SM logs, it looks like the problem occurred when repairing `keyspace1.standard1`:
```
Sep 16 08:32:13 : {"L":"INFO","T":"2023-09-16T08:32:13.595Z","N":"repair.worker 0","M":"Repairing","keyspace":"keyspace1","table":"standard1","master":"10.4.0.73","hosts":["10.4.0.73","10.4.0.75","10.4.1.127"],"ranges":[{"StartToken":-9223372036854775808,"EndToken":9223372036854775807}],"job_id":13,"_trace_id":"V0mA2RqoQUK17cRxEz01XQ"}
...
Sep 16 08:32:43 : {"L":"INFO","T":"2023-09-16T08:32:43.596Z","N":"cluster.client","M":"HTTP retry backoff","operation":"StorageServiceRepairStatus","wait":"950.856226ms","error":"after 30s: context deadline exceeded","_trace_id":"V0mA2RqoQUK17cRxEz01XQ"}
```
It looks like node `10.4.0.73` is not responding to repair status query.
What's interesting, is that SM applied small table optimization for repairing `keyspace1.standard1`, so it tried to repair all token ranges in a single repair job (`"ranges":[{"StartToken":-9223372036854775808,"EndToken":9223372036854775807}]`).
Perhaps this is the reason why node `10.4.0.73` wasn't responding (although it should handle this situation just fine)?

SM applies small table optimization to fully replicated, smaller than 1GB tables (by default).
Was `keyspace1.standard1` smaller than 1GB at repair time? Maybe there is a bug in terms of how SM checks table size.

Or is there a chance that the cluster was going through some topology changes at repair time?

**Comment by mykaul** [comment]
What's the next step here?

**Comment by aleksbykov** [comment]
> From SM logs, it looks like the problem occurred when repairing `keyspace1.standard1`:
> SM applies small table optimization to fully replicated, smaller than 1GB tables (by default). Was `keyspace1.standard1` smaller than 1GB at repair time? Maybe there is a bug in terms of how SM checks table size.

for moment when mgmt_repair is started the size of keyspace1.standard1 table is about 1TB


> Or is there a chance that the cluster was going through some topology changes at repair time?
I checked the log, and there is now any topology operation at that moment.

The latest run of the same job did not reproduce the issue.

**Comment by Michal-Leszczynski** [comment]
So maybe we are observing something similar to this [issue](https://github.com/scylladb/scylla-enterprise/issues/3400#issuecomment-1720870937).
Maybe we have two problems:
- checking table size from `"/column_family/metrics/total_disk_space_used/{name}"` endpoint can be misleading (or SM does it in an incorrect way?)
- Even though Scylla should be able to handle a full token range by splitting it, this still overwhelms the repair master node to the point where it has large latency or even stops responding.

**Comment by Michal-Leszczynski** [comment]
We have a test `TestServiceRepairOrderIntegration` checking repair order, which is strongly connected to table sizes.
In its original form, this test wasn't passing because `"/column_family/metrics/total_disk_space_used/{name}"` didn't report correct responses shortly after inserting the data to the table. This was fixed by flushing the table after inserting the data. 

At this point I though that it only affects small tables (we test it on tables with a few MB of data), because I would expect that inserting 1TB of data would result in many data flushes (but perhaps I misunderstand the concept).

So in case we want to rerun this test without this issue, the repair should be scheduled with `--small-table-threshold 0G` so that this optimization is disabled.

It would be interesting to see, if this is indeed some problem with `"/column_family/metrics/total_disk_space_used/{name}"` endpoint and whether flushing table (e.g. with `nodetool flush`) would make this optimization work as expected.

**Comment by Michal-Leszczynski** [comment]
Let's rerun when #3591 is merged.

**Comment by Michal-Leszczynski** [comment]
@aleksbykov would it be possible to rerun this scenario with SM branch [ml/fix_small_table_optimization](https://github.com/scylladb/scylla-manager/tree/ml/fix_small_table_optimization)?

**Comment by ShlomiBalalis** [comment]
> @aleksbykov would it be possible to rerun this scenario with SM branch [ml/fix_small_table_optimization](https://github.com/scylladb/scylla-manager/tree/ml/fix_small_table_optimization)?



> @Michal-Leszczynski , i think yes with @ShlomiBalalis help, because i don't familiar how correctly catch up SM custom branch @ShlomiBalalis , could you help me with it?

Sure we can rerun with it, but we'll need a list file

**Comment by Michal-Leszczynski** [comment]
The list file:
```
downloads.scylladb.com/manager/deb/unstable/unified-deb/ml/fix_small_table_optimization/2023-09-28T15:51:31Z/scylla-manager.list
```

**Comment by karol-kokoszka** [comment]
@ShlomiBalalis @aleksbykov We have a PR for scylla manager that is exepcted to fix this issue.
It's tested against our integration tests, but we would like to validate the behavior against this particular test that is the root of this issue on github.
Can we execute the test against chosen manager branch ? Or do we need to merge the PR to master first ?

**Comment by karol-kokoszka** [comment]
UPDATE: I checked the latest runs of this nemesis execution and look that they pass even tough the optimization is applied.
https://jenkins.scylladb.com/job/scylla-master/job/scylla-master-perf-regression-latency-650gb-with-nemesis/49/
There is `merge_ranges: true` saying that optimization is applied.

logs:
```
2023-10-01T13:29:17+00:00 perf-latency-nemesis-ubuntu-monitor-node-22a0b6f0-1     !INFO | scylla-manager[31250]: {"L":"INFO","T":"2023-10-01T13:29:17.179Z","N":"repair","M":"Repair order","order":9,"keyspace":"keyspace1","table":"standard1","merge_ranges":true,"_trace_id":"QJBLvUKxRpC8VUigY6Gqsw"}
...
2023-10-01T16:42:39+00:00 perf-latency-nemesis-ubuntu-monitor-node-22a0b6f0-1     !INFO | scylla-manager[31250]: {"L":"INFO","T":"2023-10-01T16:42:39.546Z","N":"repair.worker 0","M":"Repair done","_trace_id":"QJBLvUKxRpC8VUigY6Gqsw"}
2023-10-01T16:42:39+00:00 perf-latency-nemesis-ubuntu-monitor-node-22a0b6f0-1     !INFO | scylla-manager[31250]: {"L":"INFO","T":"2023-10-01T16:42:39.562Z","N":"repair","M":"Progress","percent":100,"count":6912,"success":6912,"failed":0,"_trace_id":"QJBLvUKxRpC8VUigY6Gqsw"}
2023-10-01T16:42:39+00:00 perf-latency-nemesis-ubuntu-monitor-node-22a0b6f0-1     !INFO | scylla-manager[31250]: {"L":"INFO","T":"2023-10-01T16:42:39.562Z","N":"repair","M":"Close generator","_trace_id":"QJBLvUKxRpC8VUigY6Gqsw"}
2023-10-01T16:42:39+00:00 perf-latency-nemesis-ubuntu-monitor-node-22a0b6f0-1     !INFO | scylla-manager[31250]: {"L":"INFO","T":"2023-10-01T16:42:39.563Z","N":"repair.worker 0","M":"Done","_trace_id":"QJBLvUKxRpC8VUigY6Gqsw"}
2023-10-01T16:42:39+00:00 perf-latency-nemesis-ubuntu-monitor-node-22a0b6f0-1     !INFO | scylla-manager[31250]: {"L":"INFO","T":"2023-10-01T16:42:39.572Z","N":"scheduler.11e523b2","M":"Run ended","task":"repair/c1db0292-5205-486d-8c91-3316520875e6","status":"DONE","duration":"3h13m23.167552991s","_trace_id":"QJBLvUKxRpC8VUigY6Gqsw"}
```

It means that the optimization doesn't lead to the error all the time even on a big table and the test execution against the branch containing the fix may not be the best indicator if the issue is completely solved or not. 

We can just validate if the optimization is applied or not. We do that with the new integration test already https://github.com/scylladb/scylla-manager/blob/0ee980822ad41bdfc11534983ff74afe9301d02a/pkg/service/repair/service_repair_integration_test.go#L1782

**Comment by karol-kokoszka** [comment]
Small table optimization is fixed with 3.2.2 release.
The root cause of the issue reported here is the small table optimization applied to big tables.

Closing.
Please re-open if the problem will appear again with >= 3.2.2.
